### PR TITLE
[Enhancement] Support avg rollup rewrite in MV

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：
1. Support AVG rollup rewrite into Sum/Count;
2. Ignore aggregation predicates, because aggregation predicates should be rewritten after aggregation functions' rewrite and should not be pushed down into mv scan operator;
3. Change fn's type  as 1st child has changed when rollup, otherwise physical plan will still use the old arg input's type
4. Add `enable_materialized_view_single_table_view_delta_rewrite ` config to indicate whether rewrite single table query into view-delta mvs or not. 
5. Not rewrite single table query by default.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 80 / 87 (91.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1822, 1823]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|62|67|92.54%|[148, 149, 150, 151, 153]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java [hunk 1]
```diff
@@ -514,11 +539,96 @@ private CallOperator getRollupAggregate(CallOperator aggCall, ColumnRefOperator
...
         return Expr.getBuiltinFunction(fnName, argsType, Function.CompareMode.IS_IDENTICAL);
     }
+
+    // Rewrite query agg operator by rule:
+    //  - now only support rewrite avg to sum/ count
+    // TODO: maybe we can rewrite the whole query before MV's rewrite.
+    private LogicalAggregationOperator rewriteAggregationOperatorByRules(
+            ColumnRefFactory queryColumnRefFactory,
+            LogicalAggregationOperator aggregationOperator) {
+        Map<ColumnRefOperator, CallOperator> oldAggregations = aggregationOperator.getAggregations();
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:559]: what if the mv has a avg(x) column? then here you rewrite avg(x) into sum(x)/count(x)，then rewrite fails.
it is a badcase.

Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:559]: maybe here we should consider both cases.

**Review by ABingHuang — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:559]: - MV's `avg`  will never be used in `rollup`,  `rewriteAggregationOperatorByRules ` is only called in `rewriteForRollup`. This pr only supports avg rewrite in rollup situations.
- So for a non-rollup agg rewrite, what's your suggestion? Maybe I can fix it later.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 2]
```diff
@@ -907,6 +910,14 @@ public boolean getUseScanBlockCache() {
     @VarAttr(name = ENABLE_MATERIALIZED_VIEW_VIEW_DELTA_REWRITE)
     private boolean enableMaterializedViewViewDeltaRewrite = true;
 
+    //  Whether to enable view delta compensation for single table,
+    //  - try to rewrite single table query into candidate view-delta mvs if enabled which will choose
+    //      plan by cost.
+    //  - otherwise not try to write single table query by using candidate view-delta mvs which only
+    //      try to rewrite by single table mvs and is determined by rule rather than by cost.
+    @VarAttr(name = ENABLE_MATERIALIZED_VIEW_SINGLE_TABLE_VIEW_DELTA_REWRITE)
+    private boolean enableMaterializedViewSingleTableViewDeltaRewrite = false;
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:919]: We can remove some session variables if we have ensued some situations are confirmed in the later version.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java [hunk 3]
```diff
@@ -514,11 +539,96 @@ private CallOperator getRollupAggregate(CallOperator aggCall, ColumnRefOperator
...
             return newAggCall;
         }
     }
 
     private Function findArithmeticFunction(Type[] argsType, String fnName) {
         return Expr.getBuiltinFunction(fnName, argsType, Function.CompareMode.IS_IDENTICAL);
     }
+
+    // Rewrite query agg operator by rule:
+    //  - now only support rewrite avg to sum/ count
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:554]: count(disinct xx) to bitmap?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 4]
```diff
@@ -907,6 +910,14 @@ public boolean getUseScanBlockCache() {
     @VarAttr(name = ENABLE_MATERIALIZED_VIEW_VIEW_DELTA_REWRITE)
     private boolean enableMaterializedViewViewDeltaRewrite = true;
 
+    //  Whether to enable view delta compensation for single table,
+    //  - try to rewrite single table query into candidate view-delta mvs if enabled which will choose
+    //      plan by cost.
+    //  - otherwise not try to write single table query by using candidate view-delta mvs which only
+    //      try to rewrite by single table mvs and is determined by rule rather than by cost.
+    @VarAttr(name = ENABLE_MATERIALIZED_VIEW_SINGLE_TABLE_VIEW_DELTA_REWRITE)
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java]: done.

**Comment by LiShuMing** [comment]
run starrocks_fe_unittest

**Review by kangkaisen — APPROVED** [review]

**Comment by LiShuMing** [comment]
run starrocks_fe_unittest

**Review by ABingHuang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 80 / 87 (91.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|4|50.00%|[1821, 1822]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|62|67|92.54%|[148, 149, 150, 151, 153]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/CallOperator.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|2|2|100.00%|[]|

**Pull request merged**